designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
+- When the current active profile changes, the tree is not refreshed, but is
+  greyed-out to indicate that it is stale. It is not minimized.
+- When a bucket or object is created or deleted, only that node is added or
+  removed from the tree, _without disturbing the current layout of the tree_.
+
+
+![](20200210-design-vscode-s3-tree.png)
+
+### Create and edit S3 buckets
+
+- Root node context menu (right-click):
+  - Create bucket
+- Bucket node context menu (right-click):
+  - Delete bucket
+  - New folder...
+  - Upload...
+  - Copy name
+  - Copy ARN
+
+### Create and edit S3 objects
+
+- S3 object node context menu (right-click):
+  - Delete object
+  - Copy name
+  - Copy ARN
+  - Copy URL
+  - Upload...
+  - Edit
+    - Downloads the object to a temp folder and opens it in a new VSCode editor tab.
+  - Download
+    - Does _not_ present a chooser.
+    - Downloads immediately to the OS default "Downloads" location (typically `~/Downloads`).
+    - Download status is given as a toaster, which contains a "Choose location..." link.
+      - User can change the download location via this link. This will be the
+        default location for future downloads.
+      - The default download location is also configurable as a global Toolkit
+        option. This is not S3-specific, it is used for all "Download"
+        experiences in the Toolkit.
+
+### Drag-drop from client to server
+
+- Drag a folder from the local system to the S3 tree:
+  - onto a bucket uploads as a new top-level folder in the bucket
+  - onto a folder uploads as a new subfolder
+- Drag a file from the local system to the S3 tree:
+  - onto a bucket uploads the file as a top-level object in the bucket
+  - onto a folder uploads the file as an object in the folder
+
+### Drag-drop server-side
+
+- User can copy objects or folders from one folder to another by drag-drop.
+- User can copy objects or folders _across regions_ by dragging from one region
+  to another.
+
+### S3 bucket properties
+
+Surface popular bucket properties such as "public accessible" and "versioning
+enabled" in a clear yet uncluttered manner.
+
+- Show icon (same as AWS web console) next to bucket name indicating whether it
+  is publicly accessible
+- Show icon or sigil next to bucket name indicating whether versioning is enabled
+
+### S3 object properties
+
+S3 objects have properties like size, last-modified-date, and previous versions
+(if versioning is enabled).
+
+- If >=1 previous versions are available for an object, show the count of
+  previous versions next to the object.
+- If >=1 previous versions are available for an object, context menu
+  (right-click) on the object exposes a "Show previous versions" menu item
+  (otherwise greyed-out).
+  - Previous versions are presented as children of the object.
+- Double-clicking an object always performs the "Edit" action, even if the
```

Comment:
* What do you propose the edit button does? 
* With absolutely no evidence to back this up, I feel like "Download" would be the idiomatic action off of a double click.

Reply:
"Edit" opens the file in the text editor.

> I feel like "Download" would be the idiomatic action

I would expect that to be the default choice *as an implementer* because it's easier to implement. I suspect users in a text editor actually want files to open in their text editor.

Reply:
Have you experimented to see what happens if you open a text object vs a binary object? We don't want to open a document full of hex codes and "weird symbols". If different experiences are necessary, the UX should be clear on what will happen.

Reply:
Images open in-IDE, unsure about videos. Binaries typically open as gibberish.

Does this also mean that saving an "edited" file would repush it to S3? I'm thinking that's not an especially useful use-case? My original thought was that "edit" would be something like editing the object properties.